### PR TITLE
TASK-40799 Clustering support for call events publishing in CometD layer

### PR DESCRIPTION
**Pull request by pnedonosko**
Added clustering support for call events publishing in CometD layer. It's build on exo cache as transport between nodes of the cluster.

### CODE REVIEW

**Review by ahamdi — DISMISSED** [review]

File: services/src/main/java/org/exoplatform/webconferencing/cometd/CometdWebConferencingService.java [hunk 1]
```diff
@@ -217,7 +238,428 @@
 
   /** The call handlers. */
   protected final ExecutorService        callHandlers;
+  
+  /** Cache of connected users. */
+  protected final ExoCache<String, UserCallProxy>  usersCache;
```
Comment (boubaker) [inline services/src/main/java/org/exoplatform/webconferencing/cometd/CometdWebConferencingService.java:245]:
```suggestion
  protected final ExoCache<String, UserCallProxy>  usersCache;
  
  protected final boolean                                            isClustered;  
```

File: services/src/main/java/org/exoplatform/webconferencing/cometd/CometdWebConferencingService.java [hunk 2]
```diff
@@ -1444,18 +2106,21 @@ public void rcLogs(final RemoteCall.Caller caller, final Object args) {
...
                                       WebConferencingService webConferencing,
                                       EXoContinuationBayeux exoBayeux,
+                                      CacheService cacheService,
                                       CallLogService callLogs) {
     this.identityRegistry = identityRegistry;
     this.organization = organization;
     this.webConferencing = webConferencing;
     this.exoBayeux = exoBayeux;
     this.callLogs = callLogs;
+    this.usersCache = cacheService.getCacheInstance(USER_CACHE_NAME);
```
Comment (boubaker) [inline services/src/main/java/org/exoplatform/webconferencing/cometd/CometdWebConferencingService.java]:
```suggestion
    this.usersCache = ExoContainer.hasProfile("cluster") ? cacheService.getCacheInstance(USER_CACHE_NAME) : null;
```

File: services/src/main/java/org/exoplatform/webconferencing/cometd/CometdWebConferencingService.java [hunk 3]
```diff
@@ -399,6 +841,10 @@ boolean removeClient(String sessionId) {
         boolean res = clients.remove(sessionId);
         if (clients.size() == 0) {
           webConferencing.removeUserCallListener(listener);
+          // Remove call proxy from the cache (for clustering channel notifications)
+          // Other nodes will unregister their listener stub in its WebConf service instance
+          UserCallProxy proxy = new UserCallProxy(listener.getUserId(), listener.getClientId());
+          usersCache.remove(proxy.getId());
```
Comment (boubaker) [inline services/src/main/java/org/exoplatform/webconferencing/cometd/CometdWebConferencingService.java]:
```suggestion
          if (usersCache != null) {
            // Remove call proxy from the cache (for clustering channel notifications)
            // Other nodes will unregister their listener stub in its WebConf service instance
            UserCallProxy proxy = new UserCallProxy(listener.getUserId(), listener.getClientId());
            usersCache.remove(proxy.getId());
          }
```

File: services/src/main/java/org/exoplatform/webconferencing/cometd/CometdWebConferencingService.java [hunk 4]
```diff
@@ -425,6 +871,10 @@ boolean addClient(String sessionId) {
         boolean res = clients.add(sessionId);
         if (wasEmpty && res) {
           webConferencing.addUserCallListener(listener);
+          // Add initial call proxy to the cache (for clustering channel notifications)
+          // Other nodes will listen on the cache onPut and register a client stub listener to their WebConf service instances
+          UserCallProxy proxy = new UserCallProxy(listener.getUserId(), listener.getClientId());
+          usersCache.put(proxy.getId(), proxy);
```
Comment (boubaker) [inline services/src/main/java/org/exoplatform/webconferencing/cometd/CometdWebConferencingService.java]:
```suggestion
          if (usersCache != null) {
            // Add initial call proxy to the cache (for clustering channel notifications)
            // Other nodes will listen on the cache onPut and register a client stub listener to their WebConf service instances
            UserCallProxy proxy = new UserCallProxy(listener.getUserId(), listener.getClientId());
            usersCache.put(proxy.getId(), proxy);
          }
```

File: services/src/main/java/org/exoplatform/webconferencing/cometd/CometdWebConferencingService.java [hunk 5]
```diff
@@ -932,6 +1592,7 @@ void onContainerError(String error) {
     @PostConstruct
     public void postConstruct() {
       bayeux.addListener(channelListener);
+      usersCache.addCacheListener(usersCacheListener);
```
Comment (boubaker) [inline services/src/main/java/org/exoplatform/webconferencing/cometd/CometdWebConferencingService.java]:
```suggestion
      if (usersCache != null) {
        usersCache.addCacheListener(usersCacheListener);
      }
```

File: services/src/main/java/org/exoplatform/webconferencing/cometd/CometdWebConferencingService.java [hunk 6]
```diff
@@ -940,6 +1601,9 @@ public void postConstruct() {
     @PreDestroy
     public void preDestroy() {
       // cleanup listeners
+      // XXX we cannot remove users cache listener, but
+      // we remove all client listener stubs - unregister in WebConf service
+      usersCacheListener.stop();
```
Comment (boubaker) [inline services/src/main/java/org/exoplatform/webconferencing/cometd/CometdWebConferencingService.java]:
```suggestion
      if (usersCache != null) {
        // XXX we cannot remove users cache listener, but
        // we remove all client listener stubs - unregister in WebConf service
        usersCacheListener.stop();
      }
```

File: webapp/src/main/webapp/WEB-INF/conf/webconferencing/configuration.xml [hunk 7]
```diff
@@ -225,5 +225,29 @@
       </init-params>
     </component-plugin>
   </external-component-plugins>
+  
+  <!-- Cache configuration -->
+  <external-component-plugins>
+    <target-component>org.exoplatform.services.cache.CacheService</target-component>
+    <component-plugin>
```
Comment (boubaker) [inline webapp/src/main/webapp/WEB-INF/conf/webconferencing/configuration.xml]:
```suggestion
    <component-plugin profiles="cluster">

```

File: services/src/main/java/org/exoplatform/webconferencing/cometd/CometdWebConferencingService.java (hunk 4 above)
Comment (pnedonosko) [reply services/src/main/java/org/exoplatform/webconferencing/cometd/CometdWebConferencingService.java]: Done

File: webapp/src/main/webapp/WEB-INF/conf/webconferencing/configuration.xml (hunk 7 above)
Comment (pnedonosko) [reply webapp/src/main/webapp/WEB-INF/conf/webconferencing/configuration.xml]: Done

File: services/src/main/java/org/exoplatform/webconferencing/cometd/CometdWebConferencingService.java (hunk 2 above)
Comment (pnedonosko) [reply services/src/main/java/org/exoplatform/webconferencing/cometd/CometdWebConferencingService.java]: Done

**Review by boubaker — APPROVED** [review]

**Pull request merged**